pkg/assets/create/creater.go

```diff
@@ -0,0 +1,190 @@
+package create
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+
+	"github.com/openshift/library-go/pkg/assets"
+)
+
+type CreateOptions struct {
+	// Filters allows to filter which files we will read from disk.
+	Filters []assets.FileInfoPredicate
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported.
+// Pass the context to indicate how much time you are willing to wait until all resources are created.
+func EnsureManifestsCreated(ctx context.Context, manifestDir string, restConfig *rest.Config, options CreateOptions) error {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	getRESTMapper := func() (meta.RESTMapper, error) {
+		gr, err := restmapper.GetAPIGroupResources(dc)
+		if err != nil {
+			return nil, err
+		}
+		return restmapper.NewDiscoveryRESTMapper(gr), nil
+	}
+
+	mapper, err := getRESTMapper()
+	if err != nil {
+		return err
+	}
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var lastCreateError error
+	err = wait.PollImmediateUntil(500*time.Millisecond, func() (bool, error) {
+		err, refresh := create(manifests, client, mapper, options)
+		if err == nil {
+			// No errors means all resources were created successfully, reset lastCreateError and exit the loop.
+			lastCreateError = nil
+			return true, nil
+		}
+		// We got rest-mapper error, force to refresh discovery so next create use updated mapper.
+		if refresh {
+			mapper, err = getRESTMapper()
+			if err != nil {
+				return false, err
+			}
+		}
+		lastCreateError = err
+		return false, nil
+	}, ctx.Done())
+
+	// Return the last observed set of errors from the create process instead of timeout error.
+	if lastCreateError != nil {
+		return lastCreateError
+	}
+
+	return err
+}
+
+func create(manifests map[string]*unstructured.Unstructured, client dynamic.Interface, mapper meta.RESTMapper, options CreateOptions) (error, bool) {
```

Comment:
This should take a `Context` argument so you can cancel mid-`manifests`.  When you have a large map, callers may not want to wait until the end of `manifests` before we check their context for cancellation.

Reply:
A larger map to me is 1000+ keys (i don't think we will ever have that case ;-) I think we want to cancel when we finish processing entire map (all manifests) and hit the timeout. I think in that case, we will return error anyway so the outcome will be error.

Reply:
> A larger map to me is 1000+ keys (i don't think we will ever have that case ;-)...

How fast do you expect a single `Resource(...)...Create(...)` to be?  1000+ keys each going through that seems like a big chunk of work to not allow a cancel before you finish all the keys.

Reply:
The create will be hundred of milliseconds and I don't expect us to ever create 1000+ files trough this. The poller in `PollImmediateUntil` spawns a go routine and terminate it when the timeout is reached anyway, so I don't think we want to add some complex logic here and wrap every create call with its own timeout or portion the timeout for each create call?

Reply:
> The create will be hundred of milliseconds and I don't expect us to ever create 1000+ files trough this.

100 milliseconds seems overly optimistic to me.  What if the user is calling this from some location far from their cluster?  And I think mid-`create` cancels [are fairly straightforward][1], so I don't see a need to make assumptions about input manifest counts.

[1]: https://github.com/openshift/library-go/pull/220#discussion_r253639228

Reply:
see https://github.com/openshift/library-go/pull/220#discussion_r253787937